diff --git a/Libs/vtkTeem/vtkTeemNRRDReader.h b/Libs/vtkTeem/vtkTeemNRRDReader.h
index 6864a3a870..5f268fc919 100644
--- a/Libs/vtkTeem/vtkTeemNRRDReader.h
+++ b/Libs/vtkTeem/vtkTeemNRRDReader.h
@@ -44,7 +44,6 @@
 #include <vtkSmartPointer.h>
 #include <vtkVersion.h>
 
-#include "teem/nrrd.h"
 
 /// \brief Reads Nearly Raw Raster Data files.
 ///
@@ -142,56 +141,9 @@ public:
   vtkSetMacro(DataArrayName, std::string);
   vtkGetMacro(DataArrayName, std::string);
 
-  int NrrdToVTKScalarType(const int nrrdPixelType) const
-  {
-    switch (nrrdPixelType)
-    {
-      default:
-      case nrrdTypeDefault: return VTK_VOID; break;
-      case nrrdTypeChar: return VTK_CHAR; break;
-      case nrrdTypeUChar: return VTK_UNSIGNED_CHAR; break;
-      case nrrdTypeShort: return VTK_SHORT; break;
-      case nrrdTypeUShort:
-        return VTK_UNSIGNED_SHORT;
-        break;
-        ///    case nrrdTypeLLong:
-        ///      return LONG ;
-        ///      break;
-        ///    case nrrdTypeULong:
-        ///      return ULONG;
-        ///      break;
-      case nrrdTypeInt: return VTK_INT; break;
-      case nrrdTypeUInt: return VTK_UNSIGNED_INT; break;
-      case nrrdTypeFloat: return VTK_FLOAT; break;
-      case nrrdTypeDouble: return VTK_DOUBLE; break;
-      case nrrdTypeBlock: return -1; break;
-    }
-  }
-
-  int VTKToNrrdPixelType(const int vtkPixelType) const
-  {
-    switch (vtkPixelType)
-    {
-      default:
-      case VTK_VOID: return nrrdTypeDefault; break;
-      case VTK_CHAR: return nrrdTypeChar; break;
-      case VTK_UNSIGNED_CHAR: return nrrdTypeUChar; break;
-      case VTK_SHORT: return nrrdTypeShort; break;
-      case VTK_UNSIGNED_SHORT:
-        return nrrdTypeUShort;
-        break;
-        ///    case nrrdTypeLLong:
-        ///      return LONG ;
-        ///      break;
-        ///    case nrrdTypeULong:
-        ///      return ULONG;
-        ///      break;
-      case VTK_INT: return nrrdTypeInt; break;
-      case VTK_UNSIGNED_INT: return nrrdTypeUInt; break;
-      case VTK_FLOAT: return nrrdTypeFloat; break;
-      case VTK_DOUBLE: return nrrdTypeDouble; break;
-    }
-  }
+  int NrrdToVTKScalarType(const int nrrdPixelType) const;
+  int VTKToNrrdPixelType(const int vtkPixelType) const;
+
   vtkImageData* AllocateOutputData(vtkDataObject* out, vtkInformation* outInfo) override;
   void AllocateOutputData(vtkImageData* out, vtkInformation* outInfo, int* uExtent) override { Superclass::AllocateOutputData(out, outInfo, uExtent); }
   void AllocatePointData(vtkImageData* out, vtkInformation* outInfo);
@@ -200,7 +152,7 @@ protected:
   vtkTeemNRRDReader();
   ~vtkTeemNRRDReader() override;
 
-  static bool GetPointType(Nrrd* nrrdTemp, int& pointDataType, int& numOfComponents);
+  static bool GetPointType(void* nrrdTemp, int& pointDataType, int& numOfComponents);
 
   vtkSmartPointer<vtkMatrix4x4> RasToIjkMatrix;
   vtkSmartPointer<vtkMatrix4x4> MeasurementFrameMatrix;
@@ -208,7 +160,7 @@ protected:
 
   std::string CurrentFileName;
 
-  Nrrd* nrrd;
+  void* nrrd;
 
   int ReadStatus;
 
@@ -227,7 +179,7 @@ protected:
   void ExecuteInformation() override;
   void ExecuteDataWithInformation(vtkDataObject* output, vtkInformation* outInfo) override;
 
-  int tenSpaceDirectionReduce(Nrrd* nout, const Nrrd* nin, double SD[9]);
+  int tenSpaceDirectionReduce(void* nout, const void* nin, double SD[9]);
 
 private:
   vtkTeemNRRDReader(const vtkTeemNRRDReader&) = delete;
